cmd/meroxa/root/apps/deploy.go

```diff
@@ -535,56 +537,69 @@ func (d *Deploy) validateAppJSON(ctx context.Context) error {
 		return err
 	}
 
-	if d.appConfig == nil {
-		d.lang, err = turbine.GetLangFromAppJSON(ctx, d.logger, d.path)
-		if err != nil {
-			return err
-		}
-		d.configAppName, err = turbine.GetAppNameFromAppJSON(ctx, d.logger, d.path)
-		if err != nil {
-			return err
-		}
-		config, err = turbine.ReadConfigFile(d.path)
-		d.appConfig = &config
-		if err != nil {
-			return err
-		}
+	// exit early if app config is loaded
+	if d.appConfig != nil {
+		return d.validateLanguage()
+	}
 
-		if d.gitBranch, err = turbine.GetGitBranch(d.path); err != nil {
-			return err
-		}
-		d.appName = d.prepareAppName(ctx)
+	d.lang, err = turbine.GetLangFromAppJSON(ctx, d.logger, d.path)
+	if err != nil {
+		return err
+	}
+
+	d.configAppName, err = turbine.GetAppNameFromAppJSON(ctx, d.logger, d.path)
+	if err != nil {
+		return err
+	}
+	config, err = turbine.ReadConfigFile(d.path)
+	if err != nil {
+		return err
 	}
+	d.appConfig = &config
 
-	if err = d.validateLanguage(); err != nil {
+	if d.gitBranch, err = turbine.GetGitBranch(d.path); err != nil {
 		return err
 	}
+	d.appName = d.prepareAppName(ctx)
 
 	return nil
 }
 
-func (d *Deploy) getResourceCheckErrorMessage(ctx context.Context, resources []turbine.ApplicationResource) string {
-	var errStr string
-	for _, r := range resources {
-		resource, err := d.client.GetResourceByNameOrID(ctx, r.Name)
-		if err != nil {
-			if errStr != "" {
-				errStr += "; "
-			}
+func (d *Deploy) validateResources(ctx context.Context, rr []turbine.ApplicationResource) error {
+	var errs []error
 
-			if strings.Contains(err.Error(), "could not find") {
-				errStr += fmt.Sprintf("could not find resource %q", r.Name)
-			} else {
-				errStr += err.Error()
-			}
-		} else if resource.Status.State != meroxa.ResourceStateReady {
-			if errStr != "" {
-				errStr += "; "
-			}
-			errStr += fmt.Sprintf("resource %q is not ready and usable", r.Name)
+	wrapNotFound := func(name string, err error) error {
+		if strings.Contains(err.Error(), "could not find") {
+			return fmt.Errorf("could not find resource %q", name)
 		}
+		return err
 	}
-	return errStr
+
+	for _, r := range rr {
+		resource, err := d.client.GetResourceByNameOrID(ctx, r.Name)
+		// order is important
+		switch {
+		case err != nil:
+			errs = append(errs, wrapNotFound(r.Name, err))
+		case resource.Status.State != meroxa.ResourceStateReady:
+			errs = append(errs, fmt.Errorf("resource %q is not ready and usable", r.Name))
+		case d.flags.Environment != "" && resource.Environment == nil:
+			errs = append(errs, fmt.Errorf(
+				"resource %q is not in app env %q, but in common",
+				r.Name,
+				d.flags.Environment,
+			))
+		case d.flags.Environment != "" && resource.Environment.Name != d.flags.Environment:
+			errs = append(errs, fmt.Errorf(
+				"resource %q is not in app env %q, but in %q",
+				r.Name,
+				d.flags.Environment,
+				resource.Environment.Name,
+			))
+		}
```

Comment:
One more missing case is if app is getting deployed in the common env, but the resource is in another env

Reply:
Thanks will add it, good catch

Reply:
fixed in 798a4c17